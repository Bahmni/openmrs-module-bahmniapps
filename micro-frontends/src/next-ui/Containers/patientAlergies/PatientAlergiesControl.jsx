import React, { useState, useEffect } from "react";
import PropTypes from "prop-types";
import "../../../styles/carbon-conflict-fixes.scss";
import "../../../styles/carbon-theme.scss";
import "../../../styles/common.scss";
import "./patientAllergiesControl.scss";
import { AddAllergy } from "../../Components/AddAllergy/AddAllergy";
import { FormattedMessage } from "react-intl";
import {
  fetchAllergensOrReactions,
  fetchAllergiesAndReactionsForPatient
} from "../../utils/PatientAllergiesControl/AllergyControlUtils";
import { ViewAllergiesAndReactions } from "../../Components/ViewAllergiesAndReactions/ViewAllergiesAndReactions";
import { I18nProvider } from "../../Components/i18n/I18nProvider";
import { NotificationCarbon } from "bahmni-carbon-ui";

/** NOTE: for reasons known only to react2angular,
 * any functions passed in as props will be undefined at the start, even ones inside other objects
 * so you need to use the conditional operator like props.hostApi?.callback even though it is a mandatory prop
 */

const AllergenKind = {
  DRUG: "Drug",
  FOOD: "Food",
  ENVIRONMENT: "Environment",
};
export function PatientAlergiesControl(props) {
  const { hostData, appService } = props;
  const { patient, provider, activeVisit, allergyControlConceptIdMap } = hostData;

  const isAddButtonEnabled = activeVisit && activeVisit.uuid;

  const extractAllergenData = (allergenData, allergenKind) =>
    allergenData?.setMembers
      ?.filter((allergen) => allergen.display !== "Other non-coded")
      .map((allergen) => {
        return {
          name: allergen.display,
          kind: allergenKind,
          uuid: allergen.uuid,
        };
      });

  const extractReactionData = (reactionData) =>
    reactionData?.setMembers
      ?.filter((reaction) => reaction.display !== "Other non-coded")
      .map((reaction) => {
        return { name: reaction.names[0].display, uuid: reaction.uuid };
      });

  const TransformReactionData = (reactionData) => {
    const extractedReactionData = extractReactionData(reactionData, "Reaction");
    const reactions = {};

    extractedReactionData.forEach((item) => {
      reactions[item.uuid] = { name: item.name };
    });
    return reactions;
  };

  const TransformSeverityData = (severityData) => {
    const {setMembers, answers} = severityData;
    const severities = setMembers.length > 0 ? setMembers: answers;
    return severities.map((severity) =>
        ({name: severity.display, uuid: severity.uuid}));
  }

  const TransformAllergenData = (
    medicationAllergenData,
    foodAllergenData,
    environmentAllergenData
  ) => {
    const medicationAllergens = extractAllergenData(
      medicationAllergenData,
      AllergenKind.DRUG
    );
    const environmentalAllergens = extractAllergenData(
      environmentAllergenData,
      AllergenKind.ENVIRONMENT
    );
    const foodAllergens = extractAllergenData(
      foodAllergenData,
      AllergenKind.FOOD
    );

    return [
      ...medicationAllergens,
      ...environmentalAllergens,
      ...foodAllergens,
    ];
  };

  const allergiesAndReactionsForPatient = async () => {
    const allergiesAndReactions = await fetchAllergiesAndReactionsForPatient(patient.uuid);
    const allergies = allergiesAndReactions.entry;
    const allergiesData = allergies?.map((allergy) => {
      const { resource } = allergy;
      const allergen = resource.reaction[0]?.substance?.coding?.[0].display;
      const severity = resource.reaction[0].severity;
      const note = resource.note && resource.note[0].text;
      const date = new Date(resource.recordedDate);
      const provider = resource.recorder?.display;
      const reactions = resource.reaction[0]?.manifestation.map((reaction) => {
        return reaction.coding[0].display;
      });
      return {allergen, severity, reactions, note, provider, date};
    });
    allergiesData && allergiesData.sort((a, b) => b?.date - a?.date);
    const filterSeverity = (severity) =>
      allergiesData.filter((allergy) => allergy.severity === severity);
    allergiesData
      ? setAllergiesAndReactions([
          ...filterSeverity("severe"),
          ...filterSeverity("moderate"),
          ...filterSeverity("mild"),
        ])
      : setAllergiesAndReactions([]);
  };

  const [showAddAllergyPanel, setShowAddAllergyPanel] = useState(false);
  const [isLoading, setLoading] = useState(false);
  const [transformedAllergenData, setTransformedAllergenData] = useState([]);
  const [transformedReactionData, setTransformedReactionData] = useState({});
  const [transformedSeverityData, setTransformedSeverityData] = useState([]);
  const [allergiesAndReactions, setAllergiesAndReactions] = useState([]);
  const [showSuccessPopup, setShowSuccessPopup] = useState(false);
  const [showErrorPopup, setShowErrorPopup] = useState(false);

  const noAllergiesText = (
    <FormattedMessage
      id={"NO_ALLERGIES"}
      defaultMessage={"No Allergies for this patient."}
    />
  );
  const allergiesHeading = (
    <FormattedMessage id={"ALLERGIES_DISPLAY_CONTROL_HEADING"} defaultMessage={"Allergies"} />
  );
  const addButtonText = (
    <FormattedMessage id={"ADD_BUTTON_TEXT"} defaultMessage={"Add +"} />
  );
  const loadingMessage = (
    <FormattedMessage
      id={"LOADING_MESSAGE"}
      defaultMessage={"Loading... Please Wait"}
    />
  );

  const buildAllergenAndReactionsData = async () => {
    const urls = [
      allergyControlConceptIdMap.medicationAllergenUuid,
      allergyControlConceptIdMap.foodAllergenUuid,
      allergyControlConceptIdMap.environmentalAllergenUuid,
      allergyControlConceptIdMap.allergyReactionUuid,
      allergyControlConceptIdMap.allergySeverityUuid
    ];

    try {
      setLoading(true);
      const [
        medicationResponseData,
        foodResponseData,
        environmentalResponseData,
        reactionResponseData,
        severityResponseData
      ] = await Promise.all(urls.map((url) => fetchAllergensOrReactions(url)));
      const allergenData = TransformAllergenData(
        medicationResponseData,
        foodResponseData,
        environmentalResponseData
      );
      const reactionsData = TransformReactionData(reactionResponseData);

      const severityData = TransformSeverityData(severityResponseData);

      setTransformedAllergenData(allergenData);
      setTransformedReactionData(reactionsData);
      setTransformedSeverityData(severityData);
      setLoading(false);
    } catch (e) {
      console.log(e);
    } finally {
      setLoading(false);
    }
  };

  useEffect(() => {
    buildAllergenAndReactionsData();
    allergiesAndReactionsForPatient();
  }, []);

  return (
    <>
      <I18nProvider>
      {isLoading ? (
        <div>{loadingMessage}</div>
      ) : (
        <div>
          <h2 className={"section-title-next-ui"}>
            {allergiesHeading}
            {isAddButtonEnabled && (
              <div
                className={"add-button"}
                onClick={() => {
                  setShowAddAllergyPanel(true);
                }}
              >
                {addButtonText}
              </div>
            )}
          </h2>
            {allergiesAndReactions.length === 0 ?<div className={"placeholder-text"}>{noAllergiesText}</div>:
                <ViewAllergiesAndReactions allergies={allergiesAndReactions} showTextAsAbnormal={appService.getAppDescriptor().getConfigValue("showTextAsAbnormal")}/>
            }
          { showAddAllergyPanel && (
            <AddAllergy
              reaction={transformedReactionData}
              allergens={transformedAllergenData}
              severityOptions={transformedSeverityData}
              patient={patient}
              provider={provider}
              data-testid={"allergies-overlay"}
              onClose={() => {
                setShowAddAllergyPanel(false);
              }}
              onSave={async (isSaveSuccess) => {
                if(isSaveSuccess){
                  setShowSuccessPopup(true);
                  setShowAddAllergyPanel(false);
                }
                else if(isSaveSuccess === false){
                  setShowErrorPopup(true);
                }
              }}
            />
          )}
          <NotificationCarbon messageDuration={3000} onClose={()=>{setShowSuccessPopup(false); window.location.reload()}} showMessage={showSuccessPopup} kind={"success"} title={"Allergy saved successfully"} hideCloseButton={true}/>
          <NotificationCarbon messageDuration={3000} onClose={()=>{setShowErrorPopup(false);}} showMessage={showErrorPopup} kind={"error"} title={"Error saving allergy"} hideCloseButton={true}/>
        </div>
      )}
      </I18nProvider>
    </>
  );
}

PatientAlergiesControl.propTypes = {
  hostData: PropTypes.object.isRequired,
  appService: PropTypes.object.isRequired,
};
