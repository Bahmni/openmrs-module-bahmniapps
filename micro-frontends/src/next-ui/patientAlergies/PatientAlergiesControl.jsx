import React from 'react';
import PropTypes from 'prop-types';
import { Button } from 'carbon-components-react';

/** NOTE: for reasons known only to react2angular,
 * any function nested inside one of the prop objects will be undefined at first and then later load up
 * so you need to use the conditional operator like props.hostApi?.callback even though it is a mandatory prop
 */

export function PatientAlergiesControl(props) {
  return (
    <div>
      <span>Displaying alergy control from {props.hostData.name}</span>
      <Button onClick={props.hostApi?.callback}>Click for callback</Button>
    </div>
  );
}

PatientAlergiesControl.propTypes = {
  hostData: PropTypes.shape({
    name: PropTypes.string.isRequired,
  }),
  hostApi: PropTypes.shape({
    callback: PropTypes.func.isRequired,
  }),
};
