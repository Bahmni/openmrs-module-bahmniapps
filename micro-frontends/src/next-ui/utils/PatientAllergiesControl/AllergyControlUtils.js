import axios from "axios";
import { BAHMNI_ENCOUNTER_URL, ENCOUNTER_TYPE_URL, FETCH_CONCEPT_URL, GET_ALLERGIES_URL, SAVE_ALLERGIES_URL } from "../../constants";
import { getLocale } from "../../Components/i18n/utils";

export const fetchAllergensOrReactions = async (conceptId) => {
  const locale = getLocale();
  const apiURL = FETCH_CONCEPT_URL.replace("{locale}", locale);

  const replaceConceptIdInapiURL = (conceptId) => {
    return apiURL.replace("{conceptUuid}", conceptId);
  };

  const allergenOrReactionURL = replaceConceptIdInapiURL(conceptId);

  try {
    const response = await axios.get(allergenOrReactionURL);
    return response.data;
  } catch (error) {
    console.error(error);
  }
};

export const fetchAllergiesAndReactionsForPatient = async (patientId) => {
  try {
    const response = await axios.get(GET_ALLERGIES_URL.replace("{patientId}", patientId));
    return response.data;
  }
  catch (error) {
    console.log(error);
  }
}

export const bahmniEncounter = async (payload) => {
  try{
    return await axios.post(BAHMNI_ENCOUNTER_URL, payload, {
      withCredentials: true,
      headers: {"Accept": "application/json", "Content-Type": "application/json"}
    });
  } catch (error) {
    console.log(error);
    return error;
  }
};

export const saveAllergiesAPICall = async (payload, patientId) => {
  const saveAllergies = SAVE_ALLERGIES_URL.replace("{patientId}", patientId);
  try {
    return await axios.post(saveAllergies, payload, {
      withCredentials: true,
      headers: { Accept: "application/json", "Content-Type": "application/json" },
    });
  } catch (error) {
    console.log(error);
    return error;
  }
}

export const getEncounterType = async (encounterType) => {
    try {
        const response = await axios.get(ENCOUNTER_TYPE_URL.replace("{encounterType}", encounterType));
        return response.data;
    }
    catch (error) {
        console.log(error);
    }
}