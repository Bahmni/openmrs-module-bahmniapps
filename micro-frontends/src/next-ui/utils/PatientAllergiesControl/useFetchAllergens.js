import axios from "axios";
import { useState, useEffect } from 'react';
import { FETCH_CONCEPT_URL } from "../../constants";
import { getLocale } from "../../Components/i18n/utils";

export const  useFetchAllergens = (conceptIdMap) => {
  console.log("inside useFetchAllergens");

  const [drugAllergenData, setDrugAllergenData] = useState([]);
  const [foodAllergenData, setFoodAllergenData] = useState([]);
  const [environmentAllergenData, setEnvironmentAllergenData] = useState([]);
  const [loading, setLoading] = useState(true);
  const locale= getLocale();
  const apiURL = FETCH_CONCEPT_URL.replace('{locale}',locale);

  const replaceConceptIdInapiURL = ( conceptId) => {
    console.log('conceptId',conceptId);
    return apiURL.replace('{conceptUuid}', conceptId);
    };

  const drugAllergenURL = replaceConceptIdInapiURL(conceptIdMap.drugAllergenUuid);
  const foodAllergenURL = replaceConceptIdInapiURL(conceptIdMap.foodAllergenUuid);
  const environmentAllergenURL = replaceConceptIdInapiURL(conceptIdMap.environmentalAllergenUuid);
 
    
  console.log('apiURL',apiURL);
  console.log('drugAllergenURL',drugAllergenURL);
  console.log('foodAllergenURL',foodAllergenURL);
  console.log('environmentAllergenURL',environmentAllergenURL);
    
  useEffect(() => {
  const fetchAllergenData = async () => {
    console.log(' inside fetchAllergenData');
  try {
    const [drugResponse, foodResponse, environmentResponse] = await Promise.all([
      axios.get(drugAllergenURL),
      axios.get(foodAllergenURL),
      axios.get(environmentAllergenURL),
    ]);

    console.log('drugResponse',drugResponse.data);
    console.log('foodResponse',foodResponse.data);
    console.log('environmentResponse',environmentResponse.data);

    setDrugAllergenData(drugResponse.data);
    setFoodAllergenData(foodResponse.data);
    setEnvironmentAllergenData(environmentResponse.data);

  } catch (error) {
console.error(error);
    
  }
  finally{
    setLoading(false);

  }
};
    fetchAllergenData();
  }, []);

  return {
    drugAllergenData,
    foodAllergenData,
    environmentAllergenData,
    loading,
  };
}
