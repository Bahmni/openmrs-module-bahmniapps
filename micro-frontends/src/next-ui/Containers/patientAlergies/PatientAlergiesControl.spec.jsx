import React from "react";
import { fireEvent, render, waitFor, screen } from "@testing-library/react";
import { PatientAlergiesControl } from "./PatientAlergiesControl";
import {fetchAllergiesAndReactionsForPatient} from "../../utils/PatientAllergiesControl/AllergyControlUtils";

const mockMedicationResponseData = {
  uuid: "100340AAAAAAAAAAAAAAAAAAAAAAAAAAAAAA",
  display: "Reference application allergic reactions",
  setMembers: [
    {
      uuid: "100341AAAAAAAAAAAAAAAAAAAAAAAAAAAAAA",
      display: "Allergic to bee stings",
      names: [
        {
          uuid: "100342AAAAAAAAAAAAAAAAAAAAAAAAAAAAAA",
          display: "Bee sting",
        },
      ],
    },
    {
      uuid: "100342AAAAAAAAAAAAAAAAAAAAAAAAAAAAAA",
      display: "Allergic to cats",
      names: [
        {
          uuid: "100342AAAAAAAAAAAAAAAAAAAAAAAAAAAAAA",
          display: "Cat",
        },
      ],
    },
    {
      uuid: "100343AAAAAAAAAAAAAAAAAAAAAAAAAAAAAA",
      display: "Allergic to dust",
      names: [
        {
          uuid: "100343AAAAAAAAAAAAAAAAAAAAAAAAAAAAAA",
          display: "Dust",
        },
      ],
    },
  ],
};
const mockAllergies = {
  resourceType: "Bundle",
  id: "bc0e5930-dd91-4fcd-92a6-62c8f87569a3",
  entry: [
    {
      resource: {
        resourceType: "AllergyIntolerance",
        id: "eb96124f-0dfc-4194-8567-e033a68b0b42",
        type: "allergy",
        category: [
          "food"
        ],
        criticality: "unable-to-assess",
        code: {
          coding: [
            {
              code: "547c5ebf-b064-4541-bf79-70a0bebd2bc5",
              display: "Eggs"
            }
          ]
        },
        recordedDate: "2023-10-04T23:27:44+05:30",
        reaction: [
          {
            substance: {
              coding: [
                {
                  code: "547c5ebf-b064-4541-bf79-70a0bebd2bc5",
                  display: "Eggs"
                }
              ]
            },
            manifestation: [
              {
                coding: [
                  {
                    code: "1171f21f-1810-496f-b8db-ea8dc9d05940",
                    display: "Mental status change"
                  }
                ]
              }
            ],
            severity: "moderate"
          }
        ]
      }
    }
  ]
}

const mockFetchAllergensOrReactions = jest
  .fn()
  .mockResolvedValue(mockMedicationResponseData);

const mockFetchAllergiesAndReactionsForPatient = jest.fn().mockResolvedValue(mockAllergies);

jest.mock("../../utils/PatientAllergiesControl/AllergyControlUtils", () => ({
  fetchAllergensOrReactions: () => mockFetchAllergensOrReactions(),
  fetchAllergiesAndReactionsForPatient: () => mockFetchAllergiesAndReactionsForPatient()
}));

jest.mock("../../Components/i18n/I18nProvider", () => ({
  I18nProvider: ({ children }) => <div>{children}</div>
}));

const testHostData = {
  patient: {
    uuid: "___patient_uuid__",
    givenName: "___patient_given_name__",
  },
  provider: {
    uuid: "provider#1",
    name: "demo provider"
  },
  activeVisit: {
    uuid: "___visit_uuid__",
    visitType: {
      uuid: "___visit_type_uuid__",
    },
  },
  allergyControlConceptIdMap: {
    medicationAllergenUuid: "drug_allergen_Uuid",
    foodAllergenUuid: "food_allergen_Uuid",
    environmentalAllergenUuid: "environmental_allergen_Uuid",
    allergyReactionUuid: "allergy_reaction_Uuid",
    allergySeverityUuid: "allergy_severity_Uuid",
  },
};
const testHostDataWithoutActiveVisit = {
  patient: {
    uuid: "___patient_uuid__",
    givenName: "___patient_given_name__",
  },
  activeVisit: null,
  allergyControlConceptIdMap: {
    medicationAllergenUuid: "drug_allergen_Uuid",
    foodAllergenUuid: "food_allergen_Uuid",
    environmentalAllergenUuid: "environmental_allergen_Uuid",
    allergyReactionUuid: "allergy_reaction_Uuid",
    allergySeverityUuid: "allergy_severity_Uuid",
  },
};
const mockAppService = {
  getAppDescriptor: () => ({
        getConfigValue: param => true
      }
  )
}

describe("PatientAlergiesControl", () => {
  it("renders loading message when isLoading is true", () => {
    render(<PatientAlergiesControl hostData={testHostData} appService={mockAppService}/>);
    expect(screen.getByText("Loading... Please Wait")).not.toBeNull();
  });

  it("renders allergies section when isLoading is false", async () => {
    render(<PatientAlergiesControl hostData={testHostData} appService={mockAppService}/>);

    await waitFor(() => {
      expect(screen.getByText("Allergies")).not.toBeNull();
    });
  });

  it("renders allergies section with Add button when active visit", async () => {
    const { container } = render(
      <PatientAlergiesControl hostData={testHostData} appService={mockAppService}/>
    );

    await waitFor(() => {
      expect(screen.getByText("Add +")).not.toBeNull();
      expect(container.querySelector(".add-button")).not.toBeNull();
    });
  });

  it("renders allergies section without Add button when it is not active visit", async () => {
    const { container } = render(
      <PatientAlergiesControl hostData={testHostDataWithoutActiveVisit} appService={mockAppService}/>
    );

    await waitFor(() => {
      expect(container.querySelector(".add-button")).toBeNull();
    });
  });

  it("should show the side panel when add button is clicked", async () => {
    const { container } = render(
      <PatientAlergiesControl hostData={testHostData} appService={mockAppService} />
    );

    await waitFor(() => {
      expect(screen.getByText("Add +")).toBeTruthy();
    });

    const addButton = container.querySelector(".add-button");
    fireEvent.click(addButton);
    expect(container.querySelector(".overlay-next-ui")).not.toBeNull();
  });

  it("should not show the side panel when Cancel button is clicked", async () => {
    const { container, getByTestId } = render(
      <PatientAlergiesControl hostData={testHostData} appService={mockAppService}/>
    );
    await waitFor(() => {
      expect(screen.getByText("Add +")).toBeTruthy();
    });

    const addButton = container.querySelector(".add-button");
    fireEvent.click(addButton);
    expect(container.querySelector(".overlay-next-ui")).not.toBeNull();

    fireEvent.click(getByTestId("cancel"));
    expect(container.querySelector(".overlay-next-ui")).toBeNull();
  });
});
