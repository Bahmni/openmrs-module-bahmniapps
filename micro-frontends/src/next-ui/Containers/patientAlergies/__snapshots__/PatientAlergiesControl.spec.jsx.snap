// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PatientAlergiesControl should render the component 1`] = `
<div>
  <div>
    <h2
      class="section-title"
    >
      Allergies
      <div
        class="add-button"
      >
        Add +
      </div>
    </h2>
    <div
      class="placeholder-text"
    >
      No Allergies for this patient.
    </div>
  </div>
</div>
`;
